### PR TITLE
use appveyor build matrix

### PR DESCRIPTION
**Pull request by enricosada**
job build time limit is 60 minutes, but we are running only one job.

we can use the [build matrix feature](http://www.appveyor.com/docs/build-configuration#build-matrix) to split test suite in multiple jobs (each with 60min quota)

>Every AppVeyor build consists of one or more jobs. A build is considered successful if all jobs are successful. A build immediately fails when any of its jobs fail.

without the env var `TESTSUITE` set, the old workflow is executed so no difference from command line.

We can discuss how to split better the matrix ( fsc/library/vs extension ) to use less time for compilation and more time for test suite. 
I think we can run the full `tests/fsharp` suite in appveyor ( really good for community )

this pr add `RunTests.cmd release` of

- coreunitportable7
- coreunitportable47
- coreunitportable78
- coreunitportabl259

### CODE REVIEW

**Comment by enricosada** [comment]
@KevinRansom if ok, i  can cleanup to compile only the test assemblies of the test suite ( for example build `portable7` only in `TESTSUITE=coreunitportable7` ). 

maybe is better to use `TARGETPROFILE` vs `TESTSUITE` too

**Comment by enricosada** [comment]
Changed a bit

I splitted the build in more profiles `all`, `net40`, `portable47`, `portable7`,`portable78`,`portable259`,`vs`,`cambridge_suite`,`qa_suite`

Each one compile the minimal required to run the tests ( so `portable78` doesnt build `core library 259` and `unittests 259` )

the `all` argument run all tests, it's the default for `appveyor-build.cmd` without arguments so locally works like before plus profiles 7,47 and 78 ( where missing )

Done like that so we can have more time for testing, maybe we can run vs ide test suite or additional tests in fsharp/fsharpqa suites
Ideally we can run all tests inside appveyor

@KevinRansom i think you have the [appveyor pro](http://www.appveyor.com/pricing) not the `Premium` right? because i see only one concurrent job

**Comment by enricosada** [comment]
not bad

|| build profile | time |
| net40 |10 min 10 sec
| portable7 |10 min 32 sec
| portable47 | 8 min 51 sec
| portable78 |10 min 21 sec
| portable259 | 10 min 10 sec
| vs | 8 min 53 sec
| cambridge_suite | 18 min 42 sec
| qa_suite | 7 min 9 sec

**Comment by forki** [comment]
so overall time went down?

**Comment by enricosada** [comment]
overral not, is not going to, but:

- i'll merge net40, portable 7, 47, 78 and 259 => < 50 min ( the proto compiler build is in common )
- try to enable vside tests
- add more tests to cambridge_suite and fsharpqa (i think full run should be ok)

**Comment by forki** [comment]
:heart:

**Comment by enricosada** [comment]
btw with this modify, locally i can do 

- `appveyor-build.cmd portable7` 
- `appveyor-build.cmd all` ( all profiles )
- `appveyor-build.cmd net40`
  
a crappy bat build script :crying_cat_face:

**Comment by enricosada** [comment]
ok, now `appveyor-build.cmd` support multiple profiles (comma separated) to run

- `appveyor-build.cmd /?` -> help on usage ( `/h` `/help` )
- `appveyor-build.cmd` (no args ) = `appveyor-build.cmd all`
- `appveyor-build.cmd portable7,vs,qa_suite` run portable7 + vs + qa_suite

used that to merge `net40` and all `portable` profile in one build job

minimal build is 7min per job (proto build etc)

about other test suite

- `tests/fsharpqa` i'll try to run full suite next commit
- `tests/fsharp` require fix in my other pr abount nunit3
- `vs ide` next pull request i'll try to run tests, meanwhile i'll merge with `net40`+`portable` to lower build duration

| profiles | time |
|-----------|-------|
| net40,portable7,portable47,portable78,portable259 | 27 min 45 sec |
| vs | 9 min 14 sec |
| cambridge_suite | 20 min 7 sec |
| qa_suite | 7 min 6 sec |

**Comment by enricosada** [comment]
Full fsharpqa suite

```
3172 tests passed (99.44%)
18 tests failed (0.56%)
0 tests were cascaded failures (0.00%)
0 tests returned "no result" (0.00%)
0 tests timed out (0.00%)
0 tests had test errors (0.00%)
========
3190 Total executed tests
 
25 tests were skipped
========
3215 Total tests (executed & skipped)
```

**Comment by dsyme** [comment]
What are the test failures?

**Comment by enricosada** [comment]
```
[00:08:35] CompilerOptions\fsc\highentropyva (CheckHighEntropyALSR - x64 highentropyva+)
[00:08:35]  -- failed
[00:08:35] CompilerOptions\fsc\highentropyva (CheckHighEntropyALSR - x86 highentropyva+)
[00:08:35]  -- failed
[00:08:35] CompilerOptions\fsc\highentropyva (CheckHighEntropyALSR - x64 highentropyva)
[00:08:35]  -- failed
[00:08:35] CompilerOptions\fsc\highentropyva (CheckHighEntropyALSR - x86 highentropyva)
[00:08:35]  -- failed
[00:08:35] CompilerOptions\fsc\highentropyva (CheckHighEntropyALSR - IA64 highentropyva)
[00:08:35]  -- failed
[00:09:30] CompilerOptions\fsc\subsystemversion (default is 4.00)
[00:09:30]  -- failed
[00:09:30] CompilerOptions\fsc\subsystemversion (explicit 4.00)
[00:09:30]  -- failed
[00:09:30] CompilerOptions\fsc\subsystemversion (explicit 5.01)
[00:09:30]  -- failed
[00:09:30] CompilerOptions\fsc\subsystemversion (explicit 6.00 - Vista)
[00:09:30]  -- failed
[00:09:30] CompilerOptions\fsc\subsystemversion (explicit 6.02 - Win8)
[00:09:30]  -- failed
[00:09:30] CompilerOptions\fsc\subsystemversion (explicit 6.01 - Win7 - winexe)
[00:09:30]  -- failed
[00:09:30] CompilerOptions\fsc\subsystemversion (explicit 6.01 - Win7 - exe)
[00:09:30]  -- failed
[00:09:30] CompilerOptions\fsc\subsystemversion (explicit 6.01 - Win7 - library)
[00:09:30]  -- failed
[00:09:30] CompilerOptions\fsc\subsystemversion (explicit 6.01 - Win7 - module)
[00:09:30]  -- failed
[00:09:30] CompilerOptions\fsc\subsystemversion (6.2)
[00:09:30]  -- failed
[00:09:30] CompilerOptions\fsc\subsystemversion (06.002)
[00:09:30]  -- failed
[00:09:30] CompilerOptions\fsc\subsystemversion (65535.65535)
[00:09:30]  -- failed
[00:15:49] Conformance\LexicalAnalysis\StringsAndCharacters (TripleQuoteStringInFSI02.fsx)
[00:15:49]  -- failed
```

i am running the suite locally, let me check if i can repro these

**Comment by enricosada** [comment]
locally works, btw next time i need to check better the [appveyor log](https://ci.appveyor.com/project/KevinRansom/visualfsharp-radou/build/0.0.1.1100/job/urwd7j72qwv0riy2#L4239)

all errors are related to link.exe

```
POSTCMD: [CheckSubsystemVersion.bat dummy.exe 6.02]
link: extra operand 'dummy.exe'
Try 'link --help' for more information.
Fail to execute the POSTCMD. FAIL
```

**Comment by enricosada** [comment]
ok found it, is using the wrong [link.exe](https://ci.appveyor.com/project/enricosada/visualfsharp/build/0.0.1.126#L6)

```
where link
C:\Program Files\Git\usr\bin\link.exe
```

instead of

`C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\bin\link.exe`

what's the best solution for this? hardcode visual studio path in bat files like [tests/fsharpqa/Source/CompilerOptions/fsc/subsystemversion/CheckSubsystemVersion.bat](https://github.com/Microsoft/visualfsharp/blob/master/tests/fsharpqa/Source/CompilerOptions/fsc/subsystemversion/CheckSubsystemVersion.bat#L13) ?

File: tests/fsharpqa/Source/CompilerOptions/fsc/highentropyva/CheckHighEntropyASLR.bat [hunk 1]
```diff
@@ -8,8 +8,11 @@ IF ERRORLEVEL 1 (
     )
 )
 
+
+SET LINK_EXE=%VS140COMNTOOLS%\..\..\VC\bin\link.exe
```
Comment (enricosada) [inline tests/fsharpqa/Source/CompilerOptions/fsc/highentropyva/CheckHighEntropyASLR.bat]: `VS140COMNTOOLS` is set by visual studio install

**Comment by enricosada** [comment]
About the failing tests:

i used `VS140COMNTOOLS` environment variable ( created by visual studio install ) to set correct path of `link.exe`, fixing all but one tests.

I disable fsharpqa test `Conformance\LexicalAnalysis\StringsAndCharacters (TripleQuoteStringInFSI02.fsx)` with `NO_CI` tag.

I dont know why it doesnt pass, i'll open an issue. 

Btw there were others `NO_CI` inside the same [env.lst](https://github.com/Microsoft/visualfsharp/blob/2d413fb940aa1677688454c50b8ec05cd3b6f78f/tests/fsharpqa/Source/Conformance/LexicalAnalysis/StringsAndCharacters/env.lst#L32) so something is not right

If the next pr build is ok, the full fsharpqa suite will run within appveyor , each commit , I think is really nice

**Comment by dsyme** [comment]
Great!!!!

**Comment by enricosada** [comment]
ok, ready to review and merge :smile: 

final stats

| BUILD_PROFILE | time |
|-----------------------|-------|
| net40,portable7,portable47,portable78,portable259,vs | 30 min 46 sec |
| cambridge_suite,smoke_only | 22 min 6 sec |
| qa_suite | 24 min 30 sec |

total: 1 hr 19 min

pls, i need that in the fsharp test suite pr to test it !! :smile: 
The full fsharp suite is ~35min + 7 base = ~42, we can run that too i think

**Comment by enricosada** [comment]
@KevinRansom if you dont want to use the build matrix, i can update this pr to use only one job like before, but i think the changes at appveyor-build.cmd are good, same functionality as before but also the various part can be run individually from command line.

**Comment by KevinRansom** [comment]
Merged into master, vs2015 and coreclr.  I had to make a tiny change to the appveyor-cmd to run with F# 4.1

**Pull request closed without merging**